### DISCUSSION TITLE
customising-asp-net-core-identity-ef-core-naming-conventions-for-postgresql/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#74] [Announcements]
# customising-asp-net-core-identity-ef-core-naming-conventions-for-postgresql/

In this post I'll describe how to configure EF Core to use snake_case for database object names instead of Camel Case, so they work better with PostgreSQL.

https://andrewlock.net/customising-asp-net-core-identity-ef-core-naming-conventions-for-postgresql/

**Comment by yulivee** [comment]
Hi Andrew,
Your article saved my a great pain in my project. As I was using .NET 5.0, the code regarding OnModelCreating had to be changed, as EF Core introduced a breaking change. For everyone interested and for future reference, here is the updated version to be used in OnModel Creating:

```
 protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            MigrateTablenamesToPostgres(modelBuilder);
           // .... more code ...
       }

private void MigrateTablenamesToPostgres(ModelBuilder builder)
        {
            foreach (var entity in builder.Model.GetEntityTypes())
            {
                // Replace table names
                entity.SetTableName(entity.GetTableName().ToSnakeCase());

                // Replace column names
                foreach (var property in entity.GetProperties())
                {
                    var columnName = property.GetColumnName(StoreObjectIdentifier.Table(property.DeclaringEntityType.GetTableName(), null));
                    property.SetColumnName(columnName.ToSnakeCase());
                }

                foreach (var key in entity.GetKeys())
                {
                    key.SetName(key.GetName().ToSnakeCase());
                }

                foreach (var key in entity.GetForeignKeys())
                {
                    key.SetConstraintName(key.GetConstraintName().ToSnakeCase());
                }

                foreach (var index in entity.GetIndexes())
                {
                    index.SetDatabaseName(index.GetDatabaseName().ToSnakeCase());
                }
            }
        }
```

**Comment by alexandis** [comment]
Hi Andrew. How to remove quotes in the generated MIGRATION SCRIPT instead of changing the CAPs? I emphasize on migration, because for the command execution the issue is resolved with DbCommandInterceptor. But here it is not applicable for some reason....

**Comment by kamransaleem** [comment]
Massive help for me too, thanks for detailing this so well.

The .NET 7 code was as per @yulivee comment below.

**Comment by juliancarrivick** [comment]
I hit another snag using this method with owned types. When the class name of the owned type was lexographically earlier than the owning class I would get an error when generating migrations along the lines of `The keys {'Id'} on 'OwningTable' and {'OwningTableId'} on 'OwnedTable' are both mapped to 'owning_table.pk_owning_table', but with different columns ({'id'} and {'owning_table_id'}).`

After diving into the EFCore code, I eventually realised it was because it was trying to find the primary key of the owned type _before_ the owning table had had it's name updated. I ended up resolving this by updating all table names prior to doing any properties:

``` c#
            foreach (var entity in builder.Model.GetEntityTypes())
            {
                // Replace ALL table names first, to make sure all references in the properties
                // are correct (otherwise we hit strange errors due to the alphanumeric ordering
                // of our class names!!)
                entity.SetTableName(entity.GetTableName().ToSnakeCase());
            }

            foreach (var entity in builder.Model.GetEntityTypes())
            {
                // Replace column names
                ...
            }
```

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 07 2018, 08:21, in reply to [Alex](https://disqus.com/by/disqus_IvwpFOx7XD/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/74#discussioncomment-6042751)</em>

---

<p>A good question - I sometimes create "schema" classes that I reference from the SQL statements, and use nameof(EntityName).ToSnakeCase() in those classes. I'll write a blog post explaining the approach in a bit more detail soon :)</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 14 2018, 12:52, in reply to [Alex](https://disqus.com/by/disqus_IvwpFOx7XD/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/74#discussioncomment-6042751)</em>

---

<p>Ad promised, here's the post: <a href="https://andrewlock.net/using-snake-case-column-names-with-dapper-and-postgresql/" rel="nofollow noopener" title="https://andrewlock.net/using-snake-case-column-names-with-dapper-and-postgresql/">https://andrewlock.net/using-snake-case-column-names-with-dapper-and-postgresql/</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 07 2019, 07:04, in reply to [akka nex](https://disqus.com/by/akkanex/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/74#discussioncomment-6042767)</em>

---

<p>Good spot, thanks! I've updated the post</p>

[a comment by an automated account is not shown]

**Reply by TXRock** [reply]
Great for regular POCOs, but does not work for default ASP.NET Core Identity tables when using `AspNetCore.Identity`
